standardization/src/main/scala/za/co/absa/enceladus/standardization/interpreter/StandardizationInterpreter.scala

```diff
@@ -17,109 +17,129 @@ package za.co.absa.enceladus.standardization.interpreter
+    val stdWithCollectedTemporaryColumns =
+      collectAllTempErrorColumns(stdWithErrColPreserved, errorColsWithErrColPreserved)
+    logger.info(s"Step 3.3: Drop all temporary error columns")
+    val stdWithTempErrorColumnsDropped =
+      dropAllTemporaryErrorColumn(stdWithCollectedTemporaryColumns, errorColsWithErrColPreserved)
+    logger.info(s"Step 3.4: Clean the final error column")
+    val cleanedStd = cleanTheFinalErrorColumn(stdWithTempErrorColumnsDropped, errorColsWithErrColPreserved)
+    logger.info(s"Standardization process finished, returning to the application...")
+    cleanedStd
   }
 
-  /**
-   * Perform the standardization of the dataframe given the expected schema
-   * @param df Dataframe to be standardized
-   * @param expSchema The schema for the df to be standardized into
-   */
-  def standardize(df: Dataset[Row], expSchema: StructType, inputType: String)(implicit spark: SparkSession, udfLib: UDFLibrary): Dataset[Row] = {
-    import spark.implicits._
+  private def stdApplyLogic(stdOutput: ParseOutput, field: StructField, nextErrorColumnIndex: Long)
+                           (implicit spark: SparkSession, udfLib: UDFLibrary): (Column, Column, String) = {
+    val ParseOutput(stdCol, errs) = stdOutput
+    val errField = s"error_${unpath(field.name)}_$nextErrorColumnIndex"
+    // If the meta data value sourcecolumn is set override the field name
+    val fieldName = SchemaUtils.getFieldNameOverriddenByMetadata(field)
+    (errs.as(errField),stdCol as fieldName, errField)
+  }
 
-    errorCols = ListBuffer[String]()
 
-    // Step 1 Schema validation against self inconsistencies
-    logger.info(s"Step 1: Schema validation")
+  private def validateSchemaAgainstSelfInconsistencies(expSchema: StructType): Unit = {
     val validationErrors = SchemaChecker.validateSchemaAndLog(expSchema)
     if (validationErrors._1.nonEmpty) {
       throw new ValidationException("A fatal schema validation error occurred.", validationErrors._1)
     }
+  }
 
-    // TODO: remove when spark-xml handles empty arrays
-    val df1: Dataset[Row] = if (inputType.toLowerCase() == "xml") {
-      df.select(expSchema.fields.map { field: StructField =>
-        SparkXMLHack.hack(field, "", df).as(field.name)
-      }: _*)
-    } else df
+  private def standardizeTheDataset(df: Dataset[Row], expSchema: StructType)
+                                   (implicit spark: SparkSession, udfLib: UDFLibrary): (DataFrame, ErrorCols)  = {
 
-    logger.info(s"Step 2: Standardization")
-    // step 2 - standardize
-    val std: DataFrame = df1.select(expSchema.fields.flatMap { field: StructField =>
+    val(cols, errorColNames, _) = expSchema.fields.foldLeft(List.empty[Column], List.empty[String], 1L) { (acc, field) =>
       logger.info(s"Standardizing field: ${field.name}")
+      val (accCols, accErrorCols, nextErrorColumnIndex) = acc
       if (field.name == ErrorMessage.errorColumnName) {
```

Comment:
is it necessary to expect an `ErrorMessage.errorColumnName` here? I don't expect it is going to be defined in the expected schema.

Reply:
This is a deeper problem and will be resolved as part of #421